# Fundamentals of Data Analysis 2021
---
### Sam Tracey
### December 2021
### Analysis of CAO points 2019 / 2020 / 2021
---

In [1]:
# Import Necessary Python Libraries.

# Regular Expressions
import re
# Convenient HTTP Requests
import requests as rq
import csv
# Working with dataframes and data
import pandas as pd
# Efficient working with datetimes
import datetime as dt

import numpy as np
# Extracting.pdf tabular data from websites
import tabula


<br>

# Import CAO 2021 Points

Reference: http://www.cao.ie/index.php?page=points&p=2021

***

In [2]:
# Retrieve CAO points URL.
resp = rq.get('http://www2.cao.ie/points/l8.php')

<br>

## Save Original CAO 2021 Data Set.

***


In [3]:
# Get The Current Date and Time
now = dt.datetime.now()

# Format as a string
nowstr =  now.strftime('%Y%m%d_%H%M%S')

In [4]:
# Create a File Path for the Original Data
path = 'data/cao2021_' + nowstr + '.html'

In [5]:
# Server is using the incorrect encoding, we need to fix it.
original_encoding = resp.encoding
# Change to CP1252
resp.encoding = 'cp1252'

In [6]:
# Save the Original html file
with open(path, 'w') as f:
    f.write(resp.text)

<br>

## Use Regular Expressions to Select Correrct Lines

***

In [7]:
#Compile Regular Expression for Matching Lines.
# Original Regular expression after week 4 videos
#re_course = re.compile('([A-Z]{2}[0-9]{3})  (.*?) (\#?|([0-9]{4}|[0-9]{3})|\*?)  (.*?)')
# Modified regular expression attempting to properly separate course details
# re_course = re.compile(r'([A-Z]{2}[0-9]{3})  (.*?)([0-9]{3,5})(\*?) *')
# Third version of Regular Expression as I was missing course with no points.
re_course = re.compile(r'([A-Z]{2}[0-9]{3})(.*)')


<br>

## Loop Through the Lines of the Response and Write to .csv file

***
 

In [8]:
# This was the code I originally worked on to use REGEX to extract the data.
# I really thought it was working only to realise that it ignored any courses with 
# no points allocated in either EOS or MID.


# Define path in which to save .csv file.
# path = 'data/cao2021_csv_' + nowstr + '.csv'
# Open the csv file for writing in to.
# with open(path,'w', encoding='utf-8') as file:
    # Loop through the lines of responses.
    # for line in resp.iter_lines():
        # Match only the lines we want - those representing courses
        # if re_course.fullmatch(line.decode('cp1252')):
            # Add comma delimiters after each grouping and decode line (using incorrect decoding!)
            # csv_ver = re_course.sub(r'\1, \2, \3, \4', line.decode('cp1252'))
            # csv_ver = ' '.join(csv_ver.split())
            # csv_ver = re.sub('[#*]', '', csv_ver)
            # file.write(csv_ver + '\n')
            

In [9]:
# Noticed that my original code above was missing courses with no points

# Define path in which to save .csv file.
path2021 = 'data/cao2021_csv_' + nowstr + '.csv'
# Open the csv file for writing in to.
with open(path2021,'w') as file:
    # Loop through the lines of responses.
    for line in resp.iter_lines():
        # Decode the line using cp1252
        dline = line.decode('cp1252')
        # Match only the lines we want - those representing courses
        if re_course.fullmatch(dline):
            # Define Course Code
            course_code = dline[:5]
            # Define Course Title
            course_title = dline[7:58]
            # Define first points
            points1 = dline[60:66]
            # Define second points
            points2 = dline[69:75]
            # Combine all string elements into full line with comma separation
            line_join = [course_code, course_title, points1, points2]
            # Use regex.sub to replace all # and * characters with ''
            line_join = [re.sub('[#*]', '', elem) for elem in line_join]
            # Write a comma separated line to file with new line after each write.
            file.write(','.join(line_join) + '\n')

In [10]:
# Read 2021 csv to dataframe and add headers for each column
df2021 = pd.read_csv(path2021, names=['Code', 'Title', 'Points1', 'Points2'])
# Add a year column for later analysis
df2021['Year'] = 2021

In [11]:
# Have a peak at top and bottom rows of dataframe
df2021

Code                                              Title Points1 Points2  \
0    AL801  Software Design for Virtual Reality and Gaming...  300        NaN   
1    AL802  Software Design in Artificial Intelligence for...  313        NaN   
2    AL803  Software Design for Mobile Apps and Connected ...  350        NaN   
3    AL805  Computer Engineering for Network Infrastructur...  321        NaN   
4    AL810  Quantity Surveying                            ...  328        NaN   
..     ...                                                ...     ...     ...   
944  WD211  Creative Computing                            ...  270        NaN   
945  WD212  Recreation and Sport Management               ...  262        NaN   
946  WD230  Mechanical and Manufacturing Engineering      ...  230        230   
947  WD231  Early Childhood Care and Education            ...  266        NaN   
948  WD232  Business Information Systems                  ...  261     261      

     Year  
0    2021  
1    2021  
2    2021  
3    2021  
4    2021  
..    ...  
944  2021  
945  2021  
946  2021  
947  2021  
948  2021  

[949 rows x 5 columns]

<br>

## Validation 0f 2021 Data

***

To validate the data that the Python code extracted from the 2021 URL I manually copy and pasted the data from the website into a Notepad++ page. I manually removed all blank lines and institution headers to leave only lines corresponding to courses.

This text file has been saved [here](http://localhost:8888/doc/tree/data/2021_validation.txt)

You can see that the validation textfile contain 949 lines of courses which matches the number of lines returned in the .csv file produced by the Python code above.

<br>

## Reading 2020 CAO Points From Messy Excel File


Reference: http://www.cao.ie/index.php?page=points&p=2020&bb=points
***

In [12]:
# Define Path for writing Data
path2020 = 'data/cao2020_csv_' + nowstr + '.csv'

In [13]:
# Define url to read data from
Cao2020_Url = 'http://www2.cao.ie/points/CAOPointsCharts2020.xlsx'

In [14]:
# Save an original version of the 2020 CAO Excel File directly from URL
# reference https://stackoverflow.com/questions/31126596/saving-response-from-requests-to-file
resp = rq.get(Cao2020_Url)
output = open(path2020, 'wb')
with open(path2020, 'wb') as output:
    output.write(resp.content)

In [15]:
# Read 2020 CAO points from .xslx URL
df2020 = pd.read_excel(Cao2020_Url,
                   sheet_name='PointsCharts2020_V2',
                   skiprows=range(10),
                   usecols = "A:O",
                   index_col=None)
# Add a year column for later analysis
df2020['Year'] = 2020


In [16]:
# Select only lvl 8 courses
df2020 = df2020[df2020['LEVEL'] > 7]
# Have a peak at top and bottom rows of dataframe
df2020

CATEGORY (i.e.ISCED description)  \
0                           Business and administration   
1                         Humanities (except languages)   
2                                                  Arts   
3                                                  Arts   
4                                                  Arts   
...                                                 ...   
1455                                               Arts   
1460  Information and Communication Technologies (ICTs)   
1461  Information and Communication Technologies (ICTs)   
1462                                  Personal services   
1463                 Engineering and engineering trades   

                                           COURSE TITLE COURSE CODE2  \
0                                International Business        AC120   
1                                          Liberal Arts        AC137   
2      First Year Art & Design (Common Entry,portfolio)        AD101   
3     Graphic Design and Moving Image Design (portfo...        AD102   
4     Textile & Surface Design and Jewellery & Objec...        AD103   
...                                                 ...          ...   
1455                                     Arts (options)        WD200   
1460                       Software Systems Development        WD210   
1461                                 Creative Computing        WD211   
1462                    Recreation and Sport Management        WD212   
1463           Mechanical and Manufacturing Engineering        WD230   

     R1 POINTS R1 Random * R2 POINTS R2 Random*       EOS EOS Random *  \
0          209         NaN       NaN        NaN       209          NaN   
1          252         NaN       NaN        NaN       252          NaN   
2     #+matric         NaN       NaN        NaN  #+matric          NaN   
3     #+matric         NaN       NaN        NaN  #+matric          NaN   
4     #+matric         NaN       NaN        NaN  #+matric          NaN   
...        ...         ...       ...        ...       ...          ...   
1455       AQA         NaN       AQA        NaN       AQA          NaN   
1460       279         NaN       NaN        NaN       279          NaN   
1461       271         NaN       NaN        NaN       271          NaN   
1462       270         NaN       NaN        NaN       270          NaN   
1463       253         NaN       NaN        NaN       253          NaN   

     EOS Mid-point  LEVEL                                 HEI  \
0              280      8                    American College   
1              270      8                    American College   
2         #+matric      8  National College of Art and Design   
3         #+matric      8  National College of Art and Design   
4         #+matric      8  National College of Art and Design   
...            ...    ...                                 ...   
1455           336      8   Waterford Institute of Technology   
1460           337      8   Waterford Institute of Technology   
1461           318      8   Waterford Institute of Technology   
1462           349      8   Waterford Institute of Technology   
1463           369      8   Waterford Institute of Technology   

     Test/Interview #  avp    v  Year  
0                 NaN  NaN  NaN  2020  
1                 NaN  NaN  NaN  2020  
2                   #  NaN  NaN  2020  
3                   #  NaN  NaN  2020  
4                   #  NaN  NaN  2020  
...               ...  ...  ...   ...  
1455              NaN  avp  NaN  2020  
1460              NaN  NaN  NaN  2020  
1461              NaN  NaN  NaN  2020  
1462              NaN  NaN  NaN  2020  
1463              NaN  NaN  NaN  2020  

[1027 rows x 16 columns]

In [17]:
# Select only Columns of Interest
df2020 = df2020.loc[:, ['COURSE CODE2', 'COURSE TITLE', 'EOS', 'EOS Mid-point', 'Year']]

# Remove all non-numeric characters that appeared in the Points columns
df2020['EOS'] = df2020['EOS'].replace('[a-z.#*+A-Z]', '',regex=True)
df2020['EOS Mid-point'] = df2020['EOS Mid-point'].replace('[a-z.#*+A-Z]', '',regex=True)

# Rename dataframe columns
df2020.columns = ['Code', 'Title', 'Points1', 'Points2', 'Year']



In [18]:
# Write dataframe to .csv file
df2020.to_csv(path2020, encoding='utf-8')

In [19]:
# Have a peak at top and bottom rows of dataframe
df2020


Code                                              Title Points1  \
0     AC120                             International Business     209   
1     AC137                                       Liberal Arts     252   
2     AD101   First Year Art & Design (Common Entry,portfolio)           
3     AD102  Graphic Design and Moving Image Design (portfo...           
4     AD103  Textile & Surface Design and Jewellery & Objec...           
...     ...                                                ...     ...   
1455  WD200                                     Arts (options)           
1460  WD210                       Software Systems Development     279   
1461  WD211                                 Creative Computing     271   
1462  WD212                    Recreation and Sport Management     270   
1463  WD230           Mechanical and Manufacturing Engineering     253   

     Points2  Year  
0        280  2020  
1        270  2020  
2             2020  
3             2020  
4             2020  
...      ...   ...  
1455     336  2020  
1460     337  2020  
1461     318  2020  
1462     349  2020  
1463     369  2020  

[1027 rows x 5 columns]

In [20]:
# Spot check to ensure that dataframe row 540 matches Excel row 542 (Difference in 2 = Python Zero indexing and Excel Header)
df2020.iloc[540]

Code                                        LC294
Title      Business Studies with Event Management
Points1                                       279
Points2                                       347
Year                                         2020
Name: 759, dtype: object

<br>

## Validation 0f 2020 Data

***

To validate the data that the Python code extracted from the 2020 URL I downloaded the Excel file for all CAO points from the website. I manually removed the top 10 rows which contained information pertaining to the courses but no actual course data. I then used the filter on the LEVEL column to only show the level 8 courses.This left 1028 rows which matches the number of rows returned using the Python code above.

This text file has been saved [here](http://localhost:8888/doc/tree/data/2020_Validation.xlsx)





<br>

## Reading 2019 CAO Points From PDF File
Reference: http://www.cao.ie/index.php?page=points&p=2020

***



In [21]:
# Define path for writing 2019 csv file
path2019 = 'data/cao2019_csv_' + nowstr + '.csv'

In [22]:
# Define path for writing 2019 original pdf file
path_pdf = 'data/cao2019_pdf_' + nowstr + '.pdf'

In [23]:
# Define url link for 2019 CAO points
cao2019_url = 'http://www2.cao.ie/points/lvl8_19.pdf'

In [24]:
# Save an original version of the 2019 CAO pdf File directly from URL
# reference https://stackoverflow.com/questions/31126596/saving-response-from-requests-to-file
resp = rq.get(cao2019_url)
with open(path_pdf, 'wb') as output:
    output.write(resp.content)

In [25]:
# Read 2019 data in pdf format from url and convert to .csv file saving it in specified path
# reference: https://tabula-py.readthedocs.io/en/latest/getting_started.html
# tabula.convert_into(cao2019_url, path, output_format="csv", pages='all')

In [26]:
# read pdf from URL into a list class (pdf_list) then convert to dataframe
# reference: https://towardsdatascience.com/how-to-extract-tables-from-pdf-using-python-pandas-and-tabula-py-c65e43bd754
pdf_list = tabula.read_pdf(cao2019_url, lattice=True, pages='all',output_format='dataframe')
df2019=pdf_list[0]

In [27]:
# Drop # and * symbols from EOS and Mid columns
#['EOS'] =  df2019['EOS'].str.replace(r"[*#]",'')
#df2019['Mid'] =  df2019['Mid'].str.replace(r"[*#]",'')

df2019['EOS'] = df2019['EOS'].replace('[a-z.#*+A-Z]', '',regex=True)
df2019['Mid'] = df2019['Mid'].replace('[a-z.#*+A-Z]', '',regex=True)


In [28]:
# Dropping insitution header lines
# Reference: https://stackoverflow.com/questions/29314033/drop-rows-containing-empty-cells-from-a-pandas-dataframe
# Convert fields that are '' to Numpy NaN values in Course Code column
df2019['Course Code'].replace('', np.nan, inplace=True)
# Drop all lines where Course Code column contains a Nan
df2019.dropna(subset=['Course Code'], inplace=True)
# Add a year Column for later anaylsis
df2019.columns = ['Code', 'Title', 'Points1', 'Points2']
df2019['Year'] = 2019


In [29]:
# Have a peak at top and bottom rows of dataframe
df2019

Code                                              Title Points1 Points2  \
1    AL801    Software Design with Virtual Reality and Gaming     304     328   
2    AL802               Software Design with Cloud Computing     301     306   
3    AL803  Software Design with Mobile Apps and Connected...     309     337   
4    AL805        Network Management and Cloud Infrastructure     329     442   
5    AL810                                 Quantity Surveying     307     349   
..     ...                                                ...     ...     ...   
960  WD200                                     Arts (options)     221     296   
961  WD210                       Software Systems Development     271     329   
962  WD211                                 Creative Computing     275     322   
963  WD212                    Recreation and Sport Management     274     311   
964  WD230           Mechanical and Manufacturing Engineering     273     348   

     Year  
1    2019  
2    2019  
3    2019  
4    2019  
5    2019  
..    ...  
960  2019  
961  2019  
962  2019  
963  2019  
964  2019  

[930 rows x 5 columns]

In [30]:
# Write dataframe to .csv file
df2019.to_csv(path2019, encoding='utf-8', index=False)

<br>

## Validation of 2019 Data

***

To validate the data that the Python code extracted from the 2019 URL I manually downloaded the .pdf data from the website and used Adobe Acrobat Pro 2017 to export the .pdf file to an Excel file . I deleted the top 10 rows in the Excel file then applied a filter to the Course Code. I used the filter to remove all the blank cells in the Course Code column. This left me with a total of 931 rows which matches the number of rows produced using the Python code above. I then manually compared a small sample of the two files to ensure they matched.

This Excel file used for validation has been saved [here](http://localhost:8888/doc/tree/data/2019_Validation.xlsx)

***

<br>

## Concatenation of all dataframes

***

In [31]:
# Define Path for Saving Data
path_union = 'data/cao_union_csv_' + nowstr + '.csv'

In [32]:
# Concatenate all lvl 8 CAO points for years 2019, 2020 & 2021.
df_all = pd.concat([df2019, df2020, df2021])
# Sort concatenated dataframe on Code and look at top 50 rows.
df_all.sort_values(by=['Code'], inplace=True)
df_all.head(50)

Code                                              Title Points1 Points2  \
174  AC120                             International Business     234     269   
175  AC120  International Business                        ...  294        294   
0    AC120                             International Business     209     280   
176  AC137  Liberal Arts                                  ...  271        270   
175  AC137                                       Liberal Arts     252     275   
1    AC137                                       Liberal Arts     252     270   
431  AD101             First Year Art & Design (Common Entry)             550   
311  AD101  First Year Art and Design (Common Entry portfo...   554       NaN   
2    AD101   First Year Art & Design (Common Entry,portfolio)                   
432  AD102             Graphic Design and Moving Image Design             635   
312  AD102  Graphic Design and Moving Image Design (portfo...   538       NaN   
3    AD102  Graphic Design and Moving Image Design (portfo...                   
433  AD103   Textile & Surface Design and Jewellery & Objects             545   
313  AD103  Textile and Surface Design and Jewellery and O...   505       NaN   
4    AD103  Textile & Surface Design and Jewellery & Objec...                   
5    AD202  Education & Design or Fine Art (Second Level T...                   
314  AD202  Education and Design or Fine Art (Second Level...   591       NaN   
434  AD202  Education & Design or Fine Art (Second Level T...             580   
6    AD204                               Fine Art (portfolio)                   
435  AD204                                           Fine Art             600   
315  AD204  Fine Art (portfolio)                          ...   514       NaN   
7    AD211                         Fashion Design (portfolio)                   
316  AD211  Fashion Design (portfolio)                    ...   760       679   
436  AD211                                     Fashion Design             600   
8    AD212                         Product Design (portfolio)                   
317  AD212  Product Design (portfolio)                    ...   413       NaN   
437  AD212                                     Product Design             600   
438  AD215                                     Visual Culture     300     338   
318  AD215  Visual Culture                                ...  337        300   
9    AD215                                     Visual Culture     320     389   
319  AD217  Illustration (portfolio)                      ...   738       NaN   
439  AD217                                       Illustration             625   
10   AD217                           Illustration (portfolio)                   
11   AD222                     Interaction Design (portfolio)                   
320  AD222  Interaction Design (portfolio)                ...   444       NaN   
440  AD222                                 Interaction Design             657   
45   AL801    Software Design with Virtual Reality and Gaming     303     367   
1    AL801    Software Design with Virtual Reality and Gaming     304     328   
0    AL801  Software Design for Virtual Reality and Gaming...  300        NaN   
46   AL802  Software Design with Artificial Intelligence f...     332     382   
1    AL802  Software Design in Artificial Intelligence for...  313        NaN   
2    AL802               Software Design with Cloud Computing     301     306   
2    AL803  Software Design for Mobile Apps and Connected ...  350        NaN   
3    AL803  Software Design with Mobile Apps and Connected...     309     337   
47   AL803  Software Design with Mobile Apps and Connected...     337     360   
4    AL805        Network Management and Cloud Infrastructure     329     442   
3    AL805  Computer Engineering for Network Infrastructur...  321        NaN   
48   AL805   Computer Engineering with Network Infrastructure     333     360   
5    AL810                           

In [33]:
# Write dataframe to .csv file
df_all.to_csv(path_union, encoding='utf-8', index=False)

<br>

The dataframe **df_all** is a union of 2019, 2020 and 2021 without removal of duplicates.

<br>

## Removing Duplicates and Joining Dataframes on Course Code.

In [34]:
# Define Path for Saving Data
path_join = 'data/cao_join_csv_' + nowstr + '.csv'

In [35]:
# Drop all duplicates from Dataframe based on the Code column.
df_join = df_all.drop_duplicates(subset=['Code'])
df_join

Code                                              Title Points1 Points2  \
174  AC120                             International Business     234     269   
176  AC137  Liberal Arts                                  ...  271        270   
431  AD101             First Year Art & Design (Common Entry)             550   
432  AD102             Graphic Design and Moving Image Design             635   
433  AD103   Textile & Surface Design and Jewellery & Objects             545   
..     ...                                                ...     ...     ...   
962  WD211                                 Creative Computing     275     322   
963  WD212                    Recreation and Sport Management     274     311   
964  WD230           Mechanical and Manufacturing Engineering     273     348   
947  WD231  Early Childhood Care and Education            ...  266        NaN   
948  WD232  Business Information Systems                  ...  261     261      

     Year  
174  2019  
176  2021  
431  2019  
432  2019  
433  2019  
..    ...  
962  2019  
963  2019  
964  2019  
947  2021  
948  2021  

[1214 rows x 5 columns]

<br>

The resulting dataframe has 1214 rows. We verified that this is correct by opening the df_all dataframe in Excel and using the Excel "Remove Duplicates" function on the Code column. The validation data set can be found [here](http://localhost:8888/doc/tree/data/Validate_drop_duplicates.csv)

In [36]:
# Drop Points1, Points2 and Year columns from df_join dataframe
df_join = df_join.drop(['Points1', 'Points2', 'Year'], axis=1)


In [37]:
# Create new dataframe for 2019 with only Code and Points columns.
df2019_points = df2019[['Code', 'Points1', 'Points2']].rename(columns={'Points1':'Points1_2019', 'Points2':'Points2_2019'})

In [38]:
# Create new dataframe for 20202 with only Code and Points columns.
df2020_points = df2020[['Code', 'Points1', 'Points2']].rename(columns={'Points1':'Points1_2020', 'Points2':'Points2_2020'})

In [39]:
# Create new dataframe for 2021 with only Code and Points columns.
df2021_points = df2021[['Code', 'Points1', 'Points2']].rename(columns={'Points1':'Points1_2021', 'Points2':'Points2_2021'})

In [40]:
# Use Pandas merge function to hoin all tables in one line of code.
# I found this much more efficient than the method shown in the
# Joins in Panda video.
# Reference: https://stackoverflow.com/questions/23668427/pandas-three-way-joining-multiple-dataframes-on-columns

df_merge = df_join.merge(df2019_points,
                         on='Code',
                         how='left').merge(df2020_points,
                                           on='Code',
                                           how='left').merge(df2021_points,
                                                             on='Code',
                                                             how='left')

In [41]:
# Have a look at the dataframe
df_merge

Code                                              Title Points1_2019  \
0     AC120                             International Business          234   
1     AC137  Liberal Arts                                  ...          252   
2     AD101             First Year Art & Design (Common Entry)                
3     AD102             Graphic Design and Moving Image Design                
4     AD103   Textile & Surface Design and Jewellery & Objects                
...     ...                                                ...          ...   
1209  WD211                                 Creative Computing          275   
1210  WD212                    Recreation and Sport Management          274   
1211  WD230           Mechanical and Manufacturing Engineering          273   
1212  WD231  Early Childhood Care and Education            ...          NaN   
1213  WD232  Business Information Systems                  ...          NaN   

     Points2_2019 Points1_2020 Points2_2020 Points1_2021 Points2_2021  
0             269          209          280       294             294  
1             275          252          270       271             270  
2             550                                  554            NaN  
3             635                                  538            NaN  
4             545                                  505            NaN  
...           ...          ...          ...          ...          ...  
1209          322          271          318       270             NaN  
1210          311          270          349       262             NaN  
1211          348          253          369       230             230  
1212          NaN          NaN          NaN       266             NaN  
1213          NaN          NaN          NaN       261          261     

[1214 rows x 8 columns]

In [42]:
# Replace all cells with white space with NaN values.
# Reference https://stackoverflow.com/questions/13445241/replacing-blank-values-white-space-with-nan-in-pandas
df_merge.replace(r'^\s*$', np.nan, regex=True, inplace = True)

In [43]:
# Write dataframe to .csv file
df_merge.to_csv(path_join, encoding='utf-8', index=False)

In [44]:
# View data types of each column.
df_merge.dtypes


Code             object
Title            object
Points1_2019     object
Points2_2019     object
Points1_2020     object
Points2_2020    float64
Points1_2021     object
Points2_2021     object
dtype: object

In [54]:
# Convert points colums to numeric values.
# Reference: https://stackoverflow.com/questions/36814100/pandas-to-numeric-for-multiple-columns
for col in  df_merge.columns[2:]:
    df_merge[col] = pd.to_numeric(df_merge[col], errors='coerce')

In [55]:
# View data types after conversion.
df_merge.dtypes

Code             object
Title            object
Points1_2019    float64
Points2_2019    float64
Points1_2020    float64
Points2_2020    float64
Points1_2021    float64
Points2_2021    float64
dtype: object

In [56]:
# Check some statistical values for numeric columns.
df_merge.describe()

Points1_2019  Points2_2019  Points1_2020  Points2_2020  Points1_2021  \
count    905.000000    915.000000    993.000000    988.000000    923.000000   
mean     374.893923    426.414208    391.799597    448.198381    407.666306   
std      123.959108    133.142255    128.446837    129.237735    128.706224   
min       55.000000     57.000000     55.000000     57.000000     57.000000   
25%      290.000000    339.000000    300.000000    362.000000    303.000000   
50%      348.000000    397.000000    368.000000    426.500000    391.000000   
75%      440.000000    477.500000    474.000000    510.000000    499.000000   
max      979.000000   1073.000000   1088.000000   1088.000000   1028.000000   

       Points2_2021  
count    255.000000  
mean     414.749020  
std      141.693386  
min       60.000000  
25%      293.500000  
50%      424.000000  
75%      521.500000  
max      904.000000

<br>

***

The code below is how I initially joined all the dataframes together having worked through the video "Joins in Panda". I didn't like the repetitive nature of the code nor the constant issues with indexing. Therefore I used the pd.merge() function above which, to me, is cleaner and easier to understand.

***

In [45]:
# Create new dataframe for 2019 with only Code and Points columns.
# df2019_points = df2019[['Code', 'Points1', 'Points2']].rename(columns={'Points1':'Points1_2019', 'Points2':'Points2_2019'})
# Set the dataframe index to the Code column.
# df2019_points = df2019_points.set_index('Code')

In [46]:
# Perform left outer join of df20219_points on df_join by key = Code
# df_join = df_join.set_index('Code').join(df2019_points[['Points1_2019', 'Points2_2019']])

In [47]:
# Create new dataframe for 2020 with only Code and Points columns.
# df2020_points = df2020[['Code', 'Points1', 'Points2']].rename(columns={'Points1':'Points1_2020', 'Points2':'Points2_2020'})
# Set the dataframe index to the Code column.
# df2020_points = df2020_points.set_index('Code')


In [48]:
# Perform left outer join of df2020_points on df_join by key = Code
# df_join = df_join.join(df2020_points[['Points1_2020', 'Points2_2020']])

In [49]:
# Create new dataframe for 2021 with only Code and Points columns.
# df2021_points = df2021[['Code', 'Points1', 'Points2']].rename(columns={'Points1':'Points1_2021', 'Points2':'Points2_2021'})
# Set the dataframe index to the Code column.
# df2021_points = df2021_points.set_index('Code')

In [50]:
# Perform left outer join of df2021_points1 on df_join by key = Code
# df_join = df_join.join(df2021_points[['Points1_2021', 'Points2_2021']])

In [51]:
# Sort the dataframe vy Code and take a look
# df_join.sort_values(by = ['Code'], inplace=True)
# df_join

## References

[1:Real-Python_REGEX](https://realpython.com/python-web-scraping-practical-introduction/)

[2:StackOverFlow-Iter_lines](https://stackoverflow.com/questions/16870648/python-read-website-data-line-by-line-when-available)

[3:REGEX_Syntax](https://docs.python.org/3/library/re.html)

[4:StackOverFlow-utf-8](https://stackoverflow.com/questions/13110629/decoding-utf-8-strings-in-python)

[5:Understanding_ISO-8859-1](https://mincong.io/2019/04/07/understanding-iso-8859-1-and-utf-8/)
